runbotics/runbotics-orchestrator-ui/src/src-app/store/slices/ProcessInstanceEvent/ProcessInstanceEvent.reducers.ts

```diff
@@ -18,3 +20,23 @@ export const reduceCrumbs = (
         index + 1
     );
 };
+
+export const updateSingleActiveLoopEvent = (state: ProcessInstanceEventState, action: PayloadAction<IProcessInstanceLoopEvent>) => {
+    if (state.all.nestedEvents[action.payload.loopId]) {
+        if (!state.all.nestedEvents[action.payload.loopId][action.payload.iterationNumber]) {
+            state.all.nestedEvents[action.payload.loopId][action.payload.iterationNumber] = [action.payload];
+            return;
+        }
+
+        if (state.all.nestedEvents[action.payload.loopId][action.payload.iterationNumber].some((loopEvent) => loopEvent.id === action.payload.id)) {
+            state.all.nestedEvents[action.payload.loopId][action.payload.iterationNumber] = state.all.nestedEvents[action.payload.loopId][action.payload.iterationNumber]
+                .map((loopEvent) => {
+                    if (loopEvent.id === action.payload.id) return action.payload;
+                    return loopEvent;
+                });
+            return;
+        }
+
+        state.all.nestedEvents[action.payload.loopId][action.payload.iterationNumber].push(action.payload);
```

Comment:
not sure if push will be suitable here. Messages may arrive in mixed order - if you just push it then overall panel content will be unreadable

Reply:
As it is now in my feature - messages are only pushed if there's an iteration of loop present and there wasn't any event related to the present event, when there's some event in progress and the message comes with information that it completed, the in-progress event is mapped to completed so there aren't any changes in the order of the array, there also aren't any repeats of the same event